gematria/datasets/pipelines/benchmark_cpu_scheduler.py

```diff
@@ -33,3 +36,62 @@ def setup_and_get_benchmark_core(self) -> int | None:
 
   def verify(self):
     pass
+
+
+class DefaultBenchmarkScheduler(BenchmarkScheduler):
+
+  def __init__(self):
+    self._cpu_mask = []
+
+  @staticmethod
+  def _get_neighboring_threads(cpu_index: int) -> list[int]:
+    with open(
+        f'/sys/devices/system/cpu/cpu{cpu_index}/topology/thread_siblings_list'
+    ) as thread_sibling_list_handle:
+      neighboring_threads_strings = re.split(
+          r'[-,]+', thread_sibling_list_handle.read().strip()
+      )
+      neighboring_threads = [
+          int(cpu_index_str) for cpu_index_str in neighboring_threads_strings
+      ]
+    return neighboring_threads
+
+  def _get_aux_core_and_hyperthread_pair(
+      self,
+      cpu_mask: Iterable[int],
+  ) -> tuple[int, list[int]]:
+    for cpu_index in cpu_mask:
+      neighboring_threads = self._get_neighboring_threads(cpu_index)
+      if len(neighboring_threads) != 2:
+        raise ValueError('Expected two hyperthreads per CPU.')
+
+      if (
+          neighboring_threads[0] in cpu_mask
+          and neighboring_threads[1] in cpu_mask
+      ):
+        cpus = list(cpu_mask)
+        cpus.remove(neighboring_threads[0])
+        cpus.remove(neighboring_threads[1])
+        return (cpus[0], [neighboring_threads[0], neighboring_threads[1]])
```

Comment:
I'd add one more check that `cpus` is non-empty after removing the hyperthread pair. Otherwise, this will raise a `IndexError` when there's just the hyperthread pair but no aux thread.

That said, it's probably better to invert the ifs to avoid nesting, e.g.
```py
if (
    neighboring_threads[0] not in cpu_mask
    or neighboring_threads[1] not in cpu_mask
):
  continue
cpus = list(cpu_mask)
cpus.remove(...)
cpus.remove(...)
if not cpus:
  continue
return (cpus[0], [...])
```